homeassistant/components/hdmi_cec.py

```diff
@@ -1,121 +1,290 @@
-                                   _select_device)
+    @callback
+    def _tx(call):
+        """Send CEC command."""
+        data = call.data
+        if ATTR_RAW in data:
+            command = CecCommand(data[ATTR_RAW])
         else:
-            _LOGGER.error("Failed to open adapter")
+            if ATTR_SRC in data:
+                src = data[ATTR_SRC]
+            else:
+                src = ADDR_UNREGISTERED
+            if ATTR_DST in data:
+                dst = data[ATTR_DST]
+            else:
+                dst = ADDR_BROADCAST
+            if ATTR_CMD in data:
+                cmd = data[ATTR_CMD]
+            else:
+                _LOGGER.error("Attribute 'cmd' is missing")
+                return False
+            if ATTR_ATT in data:
+                if isinstance(data[ATTR_ATT], (list,)):
+                    att = data[ATTR_ATT]
+                else:
+                    att = reduce(lambda x, y: "%s:%x" % (x, y), data[ATTR_ATT])
+            else:
+                att = ""
+            command = CecCommand(cmd, dst, src, att)
+        hdmi_network.send_command(command)
+
+    @callback
+    def _update(call):
+        """
+        Callback called when device update is needed.
+
+        - called by service, requests CEC network to update data.
+        """
+        hdmi_network.scan()
+
+    @callback
+    def _new_device(device):
+        """Called when new device is detected by HDMI network."""
+        discovery.load_platform(hass, "switch", DOMAIN,
+                                discovered={ATTR_NEW: [device]},
+                                hass_config=base_config)
+
+    def _start_cec(event):
+        """Register services and start HDMI network to watch for devices."""
+        descriptions = load_yaml_config_file(
+            os.path.join(os.path.dirname(__file__), 'services.yaml'))[DOMAIN]
+        hass.services.register(DOMAIN, SERVICE_SEND_COMMAND, _tx,
+                               descriptions[SERVICE_SEND_COMMAND])
+        hass.services.register(DOMAIN, SERVICE_VOLUME, _volume,
+                               descriptions[SERVICE_VOLUME])
+        hass.services.register(DOMAIN, SERVICE_UPDATE_DEVICES, _update,
+                               descriptions[SERVICE_UPDATE_DEVICES])
+        hdmi_network.set_new_device_callback(_new_device)
+        hdmi_network.start()
 
     hass.bus.listen_once(EVENT_HOMEASSISTANT_START, _start_cec)
+    hass.bus.listen_once(EVENT_HOMEASSISTANT_STOP, hdmi_network.stop)
     return True
+
+
+class CecDevice(Entity):
+    """Representation of a HDMI CEC device entity."""
+
+    def __init__(self, hass: HomeAssistant, device, logical):
+        """Initialize the device."""
+        self._device = device
+        self.hass = hass
+        self._icon = None
+        self._state = STATE_UNKNOWN
+        self._logical_address = logical
+        self.entity_id = "%s.%d" % (DOMAIN, self._logical_address)
+        device.set_update_callback(self._update)
```

Comment:
This also indicates the data you sent over the buss is not serializable.

Rather send non-serializable data in the `hass.data` dictionary.